velox/functions/sparksql/String.h

```diff
@@ -283,6 +286,78 @@ struct EndsWithFunction {
   }
 };
 
+/// Returns the substring from str before count occurrences of the delimiter
+/// delim. If count is positive, everything to the left of the final delimiter
+/// (counting from the left) is returned. If count is negative, everything to
+/// the right of the final delimiter (counting from the right) is returned. The
+/// function substring_index performs a case-sensitive match when searching for
+/// delim.
+template <typename T>
+struct SubstringIndexFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  // Results refer to strings in the first argument.
+  static constexpr int32_t reuse_strings_from_arg = 0;
+
+  // ASCII input always produces ASCII result.
+  static constexpr bool is_default_ascii_behavior = true;
+
+  bool isValidUTF8(const std::string& str) {
+    std::string utf8Str;
+    try {
+      utf8Str = boost::locale::conv::utf_to_utf<char>(str);
+    } catch (const boost::locale::conv::conversion_error&) {
+      return false;
+    }
+
+    return (utf8Str == str);
+  }
+
+  FOLLY_ALWAYS_INLINE void call(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>& str,
+      const arg_type<Varchar>& delim,
+      const int32_t& count) {
+    if (!(isValidUTF8(str) && isValidUTF8(delim))) {
```

Comment:
Is this behavior matches Spark?

Isn't Spark just assumes that input is valid UTF-8. Is-valid-UTF8 check is quite expensive, especially since the implementation here involves making a copy of the string.

Reply:
Yes. Spark assumes the input is valid utf-8. Already removed the is-valid-utf check.